srcs/juloo.keyboard2/Emoji.java

```diff
@@ -7,81 +7,782 @@
 import java.io.BufferedReader;
 import java.util.ArrayList;
 import java.util.HashMap;
+import java.util.List;
 
 public class Emoji
 {
-  private final String _name;
   private final KeyValue _kv;
-  private final String _desc;
 
-  private static HashMap<String, Emoji> emojis_by_name = new HashMap<String, Emoji>();
-
-  protected Emoji(String name, String bytecode, String desc)
+  protected Emoji(String bytecode)
   {
-    _name = name;
-    _kv = new KeyValue(bytecode, KeyValue.Kind.String, 0, 0);
-    _desc = desc;
-    emojis_by_name.put(name, this);
-  }
-
-  public String name()
-  {
-    return _name;
+    this._kv = new KeyValue(bytecode, KeyValue.Kind.String, 0, 0);
   }
 
   public KeyValue kv()
   {
     return _kv;
   }
 
-  public String getDescription()
+
+  private final static List<Emoji> _all = new ArrayList<>();
+  private final static List<List<Emoji>> _groups = new ArrayList<>();
+  private final static HashMap<String, Emoji> _stringMap = new HashMap<>();
+
+  public static void init(Resources res)
   {
-    return (_desc);
-  }
+    if (!_all.isEmpty())
+      return;
+
+    try
+    {
+      InputStream inputStream = res.openRawResource(R.raw.emojis);
+      BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream));
+      String line;
+
+      // Read emoji (until empty line)
+      while (!(line = reader.readLine()).isEmpty())
+      {
+        Emoji e = new Emoji(line);
+        _all.add(e);
+        _stringMap.put(line, e);
+      }
 
-  public static int num_groups = 0;
+      // Read group indices
+      if ((line = reader.readLine()) != null)
+      {
+        String[] tokens = line.split(" ");
+        for (int i = 0; i < tokens.length-1; i++)
+          _groups.add(_all.subList(Integer.parseInt(tokens[i]), Integer.parseInt(tokens[i+1])));
+      }
+    }
+    catch (IOException e) { Logs.exn("Emoji.init() failed", e); }
+  }
 
-  private static Emoji[][] emojis_by_group = new Emoji[][]{};
+  public static int getNumGroups()
+  {
+    return _groups.size();
+  }
 
-  public static Emoji getEmojiByName(String name)
+  public static Emoji[] getEmojisByGroup(int groupIndex)
   {
-    return emojis_by_name.get(name);
+    return _groups.get(groupIndex).toArray(new Emoji[0]);
```

Comment:
The List representation seems ideal for saving memory avoiding copies. Especially that this function is called for every group just to get the first emoji of each groups to make the tabs.
I feel like the rest of the code can be made to work with List with small modification, what do you think ?

Reply:
Agreed. For the initial draft I wanted to make as few changes outside of Emoji.java as possible, but I can definitely refactor the rest to work with Lists